### ISSUE TITLE
Check Method/Constructor Parameter List Wrapping

### ISSUE THREAD

**Issue by zkiss** [checkstyle/checkstyle#9118]
I would like to be able to check if the parameter list in methods and constructors is wrapped according to the following rules:

- Either all parameters are on a single line
- Or they are broken up into multiple lines, each on an individual line
- When parameters are on multiple lines, these lines are aligned

To illustrate this, here are a couple of code examples:

### Good

```
void f1(
  int p1,
  @Ann int p2,
  int p3
) {}

void f2(int p1, int p2, int p3, int p4, int p5, int p6, int p7, int p8, int... p9) {}

void f3(int p1,
        @Ann int p2,
        int p3) {}
```

### Bad

```
void f1(int p1, int p2,
        int p3) {}

void f2(int p1,
    int p2,
    int p3) {}
```

I have created a [proof of concept](https://github.com/zkiss/checkstyle/compare/master...zkiss:params_checks) for this, that allows me to perform this check. I have implemented this using multiple `Checks` that check different aspects of the problem: alignment and parameter numbers per line.

Is this something that would make sense to add? I am not sure I will have time to properly submit a PR for this, but I would be interested in some feedback regarding if this use case is something you would consider to support and if it is worth for me to put more effort in, or if I should just package these checks in an extension jar for my own use.

Thanks,
Zoltan

**Comment by pbludov** [comment]
How will this check handle long class/method names like [this](https://docs.jboss.org/hibernate/orm/4.1/javadocs/org/hibernate/service/jdbc/connections/spi/AbstractDataSourceBasedMultiTenantConnectionProviderImpl.html)?
What about
```java
class AbstractClassWithSomeLongNameThatCannotBeChangedDueToNamingPolicyImpl
    implements AbstractClassWithSomeLongNameThatCannotBeChangedDueToNamingPolicy {

    @Override
    public abstract AbstractClassWithSomeLongNameThatCannotBeChangedDueToNamingPolicyImpl someMethodWithLondNameThatCannotBeChangedToo(int firstParameter,
                                                                                                                                       int secondParameter) {}
```
}

**Comment by zkiss** [comment]
In my opinion line length is not a concern of these checks, there is a separate check for this. All the check is supposed to do is what I described above, which is enforcing the relative alignment and line wrapping of parameters.

Personally, in this specific case I would perhaps wrap it like shown below, but this is really up to code authors to decide, and the approach chosen does not affect how the requested feature would behave.

```
class AbstractClassWithSomeLongNameThatCannotBeChangedDueToNamingPolicyImpl
    implements AbstractClassWithSomeLongNameThatCannotBeChangedDueToNamingPolicy {

    @Override
    public abstract AbstractClassWithSomeLongNameThatCannotBeChangedDueToNamingPolicyImpl someMethodWithLondNameThatCannotBeChangedToo(
        int firstParameter,
        int secondParameter
    ) {}
}
```

**Comment by zkiss** [comment]
For further clarification on how I imagine the line wrapping and aligment checks to behave, what they would accept and reject, I would recommend taking a look at the tests included in the POC I mentioned above.

**Comment by romani** [comment]
@zkiss , thanks  a lot for reaching us.

Yes, we encourage contributors to create extensions for all their ideas of code validation - it is always good approach.
we can keep reference to you at https://checkstyle.org/#Related_Tools_Active_Tools 

accepting a PR with Checks is definitely not a easy step, as quality bar is very high (all CIs to pass, multiple human very detailed review, ...). 

There is https://github.com/sevntu-checkstyle/sevntu.checkstyle extension where quality bar is significantly lower (but still exists) and we keep their experimental Checks (or prone to false positives by design or checkstyle limitation,  or ....). If you think that  you have time to contribute to that project, at least it your Checks will be placed in more popular extension project, and might be supported by somebody else.

Both your ideas make sense:
AlignParameters (SameIndentationParameters)
SingleLineForAllOrIndidualLineParameters

Both of them will be discussed/accpeted separately, Checks should reasonably working separately.

No matter what target project you choose, you will do all himself, maintainers will just guide you in right direction.

**Comment by tfij** [comment]
Hi,
Any update on this issue?
How can I easily apply this check in my code?
This is exactly what I'm looking for.

**Comment by tfij** [comment]
@romani what is needed to implement/add such check to the checkstyle?

**Comment by tfij** [comment]
Thanks, @zkiss 
I've tested
https://github.com/zkiss/checkstyle-custom-checks/blob/master/src/main/java/hu/zkiss/checkstyle/MethodParameterLinesCheck.java and
https://github.com/zkiss/checkstyle-custom-checks/blob/master/src/main/java/hu/zkiss/checkstyle/MethodParameterAlignmentCheck.java
The first one is more important in my project and works fine.
For the second one I've had an error
```
Cannot invoke "com.puppycrawl.tools.checkstyle.api.DetailAST.getNextSibling()" because "r" is null
```
After a fix, there was another error and I fix it too.
Both fixes are in this PR https://github.com/zkiss/checkstyle-custom-checks/pull/1

**Can you release it as a lib in maven central?**
Maybe, eventually, it will be part of the checkstyle...
@romani do you see a chance to add it to the checkstyle in near future?

**How I test it**
I've compiled your classes and packed them into a jar file.
Then configure the maven checkstyle plugin and add jar as a dependency:
```
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-checkstyle-plugin</artifactId>
                <version>3.2.0</version>
                <configuration>
                    <configLocation>src/main/resources/checkstyle.xml</configLocation>
                    <encoding>UTF-8</encoding>
                    <consoleOutput>true</consoleOutput>
                    <failsOnError>true</failsOnError>
                    <linkXRef>false</linkXRef>
                </configuration>
                <executions>
                    <execution>
                        <id>validate</id>
                        <phase>validate</phase>
                        <goals>
                            <goal>check</goal>
                        </goals>
                    </execution>
                </executions>
                <dependencies>
                    <dependency>
                        <groupId>pl.tfij</groupId>
                        <artifactId>checkstyle</artifactId>
                        <version>1.0</version>
                        <scope>system</scope>
                        <systemPath>${project.basedir}/libs/tfijchecks-1.0-SNAPSHOT.jar</systemPath>
                    </dependency>
                </dependencies>
            </plugin>
```
Then I used it in checkstyle.xml:
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="hu.zkiss.checkstyle.checkstyle.MethodParameterAlignmentCheck"></module>
        <module name="hu.zkiss.checkstyle.checkstyle.MethodParameterLinesCheck"></module>
    </module>
</module>
```

**Comment by tfij** [comment]
@zkiss base on your code I've created lib with checks and released it to maven central. I've also added java record support. Here is my git repo with detail https://github.com/tfij/check-tfij-style.

@romani hope it could be a part of checkstyle... or you can think about changing the distribution model - separate the checkstyle engine and checks.

**Comment by konopka** [comment]
Hey folks, 
I have created implementation for this issue here: https://github.com/konopka/checkstyle/tree/9118_parameter_placement_and_alignment_checks 
If you [@romani , @pbludov ] find it OK then please approve this issue, so I can create a PR.

**Comment by romani** [comment]
>but splitting it into two separate checks expresses more clearly what is the intention of the check(s)

In this case they should be meaningful separately, completely. To let users play with composition. In examples we can show usage with other Check, to describe possible solution.

-------

You mentioned 

> void methodGoodAlignment2(int p1, 
                              int p2, int p3);

In description there is Bad alignment:

> void f1(int p1, int p2,
        int p3) {}

As for me this is same "bad"(for this Check) formatting.

Is this formatting is what you want to allow in you style?

**Comment by konopka** [comment]
@romani  yes, ParameterAlignmentCheck alone would allow `methodGoodAlignment2` case. And when you'd combine that with ParameterPlacementCheck which would require parameters to be on separate lines for example, then your good options will be one of:
```
    void methodGoodAlignmentOnSeparateLines1(
         int p1, 
         int p2,
         int p3);
```
or
```
    void methodGoodAlignmentOnSeparateLines2(int p1, 
                                             int p2,
                                             int p3);
```

Anyways, if you prefer this to be all in single check, then I can rewrite it

**Comment by romani** [comment]
@zkiss , please share feedback on this cases.

**Comment by zkiss** [comment]
Gosh, it's been a while :) I also prefer the separation of the behaviour into two checks. My reasoning is that I can imagine code styles where people want to allow multiple parameters on the same line, but also not all of the parameters being in one line.

```
void couldBeLegit(
  int one, int two,
  int three, int four, int five
) { ... }
```

I admit I don't envision using the check I called `MethodParameterLinesCheck` in isolation, ie I don't envision the following to be accepted in codebases:
```
void allInSeparateLinesButNotAligned(int first,
  int second,
     int third,
 int fourth) { ... }
```

But overall I like the modularity of the design, and also makes testing the checks simpler.

**Comment by zkiss** [comment]
I should add more explanation; in my POC code, the purpose of the check `MethodParameterLinesCheck` is to ensure that either **all** parameters in method declarations are in **separate** lines, or **all** on the **same** line.

ie
```
void allSameEvenIfLong(int p1, int p2, boolean p3, long p4, Object p5, short p6, Integer p7, int p8, int p9, int p10, int p11, int p12, int p13, int p14, Object... p15) { ... }

void allSeparateEvenIfMisaligned(int p1,
         int p2,

             int p3,
         int p4

) { ... }
```

The other check in my POC is `MethodParameterAlignmentCheck` which means to ensure that parameter declarations in methods all start on the same exact position in all lines (ie they're aligned), regardless of how many of the parameters are declared on the same line.

**Comment by romani** [comment]
> MethodParameterLinesCheck ,ParameterPlacementCheck

Role of such future check is strange.
If user placed all on one line - ok.
if user placed all on multiple lines with multiple params in line - ok
if user placed all on individual lines with single params in line - ok

so it actually can enforce multiple strategies of placements and in this case we can make separate Checks for this as combining all at one is always maintenance problem.

This issue looks like wants placement strategy to be SingleLineOrIndividualLinesParameterPlacementCheck but there a lot of other strategies ..... and it is not a  target of this issue ticket.


we discuss here something like AllParametersIntendedSameAsFirstParameterCheck , but dependency on some XxxxxPlacementCheck is very weak, as it will do all validations as intended without extra check. Name of it very specific so if some users needs different strategy, they will create new Check. Making Check named in general way and customize by parameters is usually ton of problem in long run to reconcile all user demands.

AllParametersIntendedSameAsFirstParameterCheck can has Check's parameter `singlePameterPerLine`.
```
// When AllParametersIntendedSameAsFirstParameterCheck.singlePameterPerLine = false
// no violation below
void couldBeLegit(
  int one, int two,
  int three, int four, int five
) { ... }

// no violation below
void couldBeLegit( int one, int two,
                               int three, int four, int five
) { ... }

// violation below
void couldBeLegit( int one, int two,
            int three, int four, int five
) { ... }

```

```
// When AllParametersIntendedSameAsFirstParameterCheck.singlePameterPerLine = true
// no violation below
void couldBeLegit(
  int one,
  int two,
  int three, 
  int four, 
  int five
) { ... }

// no violation below
void couldBeLegit( int one,
                               int two,
                               int three, 
                               int four, 
                               int five
) { ... }

// violation below
void couldBeLegit(
      int one,
  int two,
  int three, 
  int four, 
  int five
) { ... }

// no violation below
void couldBeLegit( int one,
                       int two,
                       int three, 
                       int four, 
                       int five
) { ... }
```

**Comment by romani** [comment]
if we notice that we losing energy in design disagreements, I highly  recommend to move Checks you propose to sevntu.checks and let them be there and serve your needs asap. It is not dead project it is official sandbox project and it has maven central deployments. You will meet me there as well.

**Comment by rnveach** [comment]
It is a dead project in that no one maintains sevntu. No one really fixes issues there. There was a time CI was failing from just lack of maintenance. If a check goes there, it is more likely to be merged as well as die in the review process, however it will most likely never be touched again. I think during one of our grammar/java updates we decided not to do any changes to sevntu. There is no metadata there, so checks probably will have issues with eclipse-cs/sonar if you need that.

**Comment by konopka** [comment]
@romani I agree that something that needs such a ridiculous long name as 'SingleLineOrIndividualLinesParameterPlacementCheck' is a bad idea.

From style perspective, what we have here are basically two concerns on how to place parameters in code:
1. horizontal direction - whether the parameters on multiple lines should be aligned on a single column
2. vertical direction - how the parameters should be distributed over multiple lines

My initial implementation, see PR, handles above concerns as two separate checks. 
But as I have mentioned previously, it can be rewritten as single check with multiple options, such as
```
ParameterPlacementCheck
    boolean alignHorizontallyWithFirstParameter;
    VerticalParameterAlignmentOption verticalAlignment;
```
where options would be something like
```
    EachParameterOnNewLine - hard requirement for every param to be on its own line
    EachParameterOnNewLineSingleLineAllowed - same as above, but if all the params fit on a single line, then allow it
    EachParameterOnSeparateLine - this would allow first parameter to be on the same line as method definition
    EachParameterOnSeparateLineSingleLineAllowed - same as above, but if all the params fit on a single line, then allow it
```

**Comment by zkiss** [comment]
I am super happy to see such a lively conversation on this issue. Before all the recent activity I certainly had no hope for this feature (these features) to make it into checkstyle.

Personally, I am happy with any solution that fully allows me to perform the checks outlined in the initial post:

> check if the parameter list (...) is wrapped according to the following rules:
> - Either all parameters are on a single line
> - Or they are broken up into multiple lines, each on an individual line
> - When parameters are on multiple lines, these lines are aligned

I think that it makes sense to implement this via multiple checks because
- it yields better configurability, which allows catering for more possible styles
- simpler check code

Since I am not a project committer, I of course have no choice other than accepting the decision of the maintainers on the implementation details. I participate in the conversation in order to clarify any points that are not clear in the feature request, and also because I am interested in the technical decisions and details.

One thing I would like to highlight is that any approach taken on this issue is additive, meaning it will not break existing code styles, and it will only give more options for people to use if they wish. There doesn't seem to be a huge community uproar so far about this particular check missing; these concerns (especially the relaxed form of parameter alignment, where the first line is ignored in case it is on the same line as the method definition) are widely supported in IDE code formatters by default and are commonly followed by convention; parameters being aligned with a double-indent on continuation lines.

With the above point in mind, I would like to refine what I said earlier wrt the check for separate lines (what I called `MethodParameterLinesCheck` in my POC): I do see the point of it being a separate check after all, as I can see it being useful to enforce the parameters being on separate lines, but not necessarily be aligned, as the first parameter might be allowed to stay on the same line the method is declared on (this is not my preference though). The parameter alignment check might get a flag to allow this or not; I would not object to this.

I had a quick look at https://github.com/checkstyle/checkstyle/pull/15773 and I actually like the names `ParameterAlignment` and `ParameterPlacement`. `ParameterPlacement` is perhaps a bit less clear, and maybe needs further refinement wrt how to cater for different styles/strategies for checking. Also, based on the name it is not immediately clear that it deals with checking the vertical direction of parameter placement.

### Reactions to more recent comments:

> so it actually can enforce multiple strategies of placements and in this case we can make separate Checks for this as combining all at one is always maintenance problem.

and

> Making Check named in general way and customize by parameters is usually ton of problem in long run to reconcile all user demands.

Agree. These two points sound very similar

> AllParametersIntendedSameAsFirstParameterCheck

- If all parameters are same as first parameter, then that includes the first parameter as well, so the name can be simplified to `AllParametersIndentedSame`. But I think `ParameterAlignment` is a strong competitor to this name - although it does not communicate clearly what it enforces, but I am not sure what else can be enforced about alignment, if not that it's the same...
- Typo in `Intended`, it is correctly written `Indented` - just want to highlight in order to avoid it getting into a final solution.

In some of the examples above I spotted a space that I would envision to flag as invalid, but they were said to be valid:

```
// no violation below
void couldBeLegit2( int one, int two,
                   int three, int four, int five
) { ... }

// no violation below
void couldBeLegit4( int one,
                   int two,
                   int three, 
                   int four, 
                   int five
) { ... }
```
I would flag both the above cases because the first parameter `int one` have a space in front of them, but the rest of the parameters on subsequent lines are aligned with the space character, not with the position where the `int` starts. I would like the alignment to start at the same position where `int one` starts, excluding the space character in front of it.

> design disagreements

Project owner wins, especially if in favour of adding the requested feature (feature != implementation).

> `VerticalParameterPlacementOption`
> ```
> EachParameterOnNewLine - hard requirement for every param to be on its own line
> EachParameterOnNewLineSingleLineAllowed - same as above, but if all the params fit on a single line, then allow it
> EachParameterOnSeparateLine - this would allow first parameter to be on the same line as method definition
> EachParameterOnSeparateLineSingleLineAllowed - same as above, but if all the params fit on a single line, then allow it
> ```
I think it would make sense to further break down this one aspect into multiple options, as this seems to be a 'matrix enum', which captures multiple aspects and types of behaviours in a single dimension. Perhaps the config could look like
```
lines:
  boolean allowAllOnSingleLine; // if false, doesn't allow
  boolean allowContinueDefinition; // if true, allows same line as method def
```

### Conclusion

Right now, there doesn't seem to be too many options after all for these two checks, so perhaps there is no need for multiple different checks, to represent the different strategies.

- ParameterAlignment
  - boolean allowContinueDefinition: allow first param to be on same line as the definition
- ParameterSeparateLines
  - boolean allowSingleLine: allow all to be on same line + definition + closing parenthesis
  - boolean allowContinueDefinition: allow first to be on same line as definition
  - boolean allowClosingParenthesis: allow last to be on same line as closing parenthesis

**Comment by romani** [comment]
>long name as 'SingleLineOrIndividualLinesParameterPlacementCheck' is a bad idea.

Unfortunately , we see huge problems with short names and mega Checks that can do a lot. We have bunch of examples, and it is hard to maintain them.

**Comment by zkiss** [comment]
Tangent, but in case there are short named mega checks which are posing a problem, isn't it possible to split them up into smaller checks with new names, keep the old name for backwards compatibility for a defined period of time and provide the backwards compatibility implementation via composition?

**Comment by romani** [comment]
Any split, will be big breaking changes for users, in most cases it will mean that Check stoped to be used , new checks are not added.
We are very demanding on names , as all names becomes kind of API, and any changes will be a pain for users for us to explain and respond. 
We just a library, and there are a lot of plugins and systems that depends on such names. We try to care about all users and all systems.
That is why I suggest to keep Checks on some extension to prove design it by time. In extensions we do not keep same compatibility and experiment easier.
